### PR TITLE
Remove regex requirement

### LINKED ISSUE #875: docs.rs build is failing
## Bug Summary

Failed docs.rs build: https://docs.rs/crate/ibc/0.45.0/builds/917198

```
[INFO] [stderr] error[E0422]: cannot find struct, variant or union type `LineColumn` in crate `proc_macro`
[INFO] [stderr]    --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/safe-proc-macro2-1.0.36/src/wrapper.rs:485:33
[INFO] [stderr]     |
[INFO] [stderr] 485 |                 let proc_macro::LineColumn { line, column } = s.start();
[INFO] [stderr]     |                                 ^^^^^^^^^^ not found in `proc_macro`
[INFO] [stderr]     |
[INFO] [stderr] help: consider importing one of these items
[INFO] [stderr]     |
[INFO] [stderr] 1   + use crate::LineColumn;
[INFO] [stderr]     |
[INFO] [stderr] 1   + use crate::fallback::LineColumn;
[INFO] [stderr]     |
[INFO] [stderr] help: if you import `LineColumn`, refer to it directly
[INFO] [stderr]     |
[INFO] [stderr] 485 -                 let proc_macro::LineColumn { line, column } = s.start();
[INFO] [stderr] 485 +                 let LineColumn { line, column } = s.start();
[INFO] [stderr]     |
[INFO] [stderr] 
```

## Details

Rust [removed](https://github.com/rust-lang/rust/pull/111571) `proc_macro::LineColumn` which is being used in `safe-proc-macro2` (dependency to `safe-regex`) under `span-locations` feature.

Probably docsrs is building the dependencies with `--all-features` which breaks the build in recent nightlies.

The fix requires an update from `safe-proc-macro2` > `safe-regex`.

## Version

`0.42.0`-`0.45.0`

### PR DESCRIPTION
**Pull request by rnbguy**
Closes: #875

## Description

- Removes any _regex_ requirement
- Removes `safe_regex` which is breaking our docsrs builds

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/applications/transfer/coin.rs [hunk 1]
```diff
@@ -59,24 +58,26 @@ where
...
-        let (m1, m2) = matcher.match_slices(coin_str.as_bytes()).ok_or_else(|| {
-            TokenTransferError::InvalidCoin {
+        // old regex: ^(?<amount>[0-9]+)(?<denom>[a-zA-Z0-9/:\\._-]+)$
+
+        let (amount, denom) = coin_str
+            .chars()
+            .position(|x| !x.is_numeric())
+            .map(|index| coin_str.split_at(index))
+            .filter(|(amount, denom)| !amount.is_empty() && !denom.is_empty())
+            .filter(|(_, denom)| {
```
Comment (soareschen) [inline crates/ibc/src/applications/transfer/coin.rs:75]: I think filter means that you will exclude the invalid characters, but the result would be valid. In the original case, invalid characters will cause the method to fail.

File: crates/ibc/src/applications/transfer/coin.rs [hunk 2]
```diff
@@ -59,24 +58,26 @@ where
...
+                !denom.contains(|x| {
+                    !matches!(x, 'a'..='z' | 'A'..='Z' | '0'..='9' | '/' | ':' | '\\' | '.' | '_' | '-')
```
Comment (soareschen) [inline crates/ibc/src/applications/transfer/coin.rs]: I think `\\` was used as an escape for `\`, which is an escape for `.` in the original regex.

**Review by soareschen — CHANGES_REQUESTED** [review]
We need to write tests to ensure that the new implementation works as expected.

**Comment by rnbguy** [comment]
> We need to write tests to ensure that the new implementation works as expected.

Yea. Tests are coming :slightly_smiling_face: I wanted to confirm the parsing logic before that.

File: crates/ibc/src/applications/transfer/coin.rs [hunk 3]
```diff
@@ -38,6 +37,19 @@ pub struct Coin<D> {
     pub amount: Amount,
 }
 
+impl<D> Coin<D> {
+    pub fn new<T, DT>(amount: T, denom: DT) -> Self
+    where
+        T: Into<Amount>,
+        DT: Into<D>,
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/coin.rs]: A note regarding these bounds, as this method would get available to our users too (not only for the tests)

Although they've made improvements in cleaning up test arguments, and at first glance, it seems to give users more flexibility, it's important to note that there are situations where both `amount` and `denom` might come from either parsing a string or using `TryFrom`. This means not all conversions can simply turn into `Amount` and `D` so users may obtain them from conversions that might fail.

In fact, we even shouldn’t use `TryFrom` or similar restrictions in this specific situation. As particularly `Coin` is generic over denoms and We can't foresee how hosts will obtain this data. So any restrictions will assume a certain approach, which may not always work. Also, e.g. in cases like [here](https://github.com/cosmos/ibc-rs/blob/1aa853bb646d37aab3848096b4221450f3a2e760/crates/ibc/src/applications/transfer/coin.rs#L94-L97) and [here](https://github.com/cosmos/ibc-rs/blob/dc8e4161dd5e00a2031cf3feb58dad1ed7e26bdb/crates/ibc/src/applications/transfer/packet.rs#L46), it is easier to work with the own types.

Btw, if still want to keep our tests args as is, Perhaps we should move this method (as a helper) under the `tests` module and only have it for the `RawCoin` type. Wdyt?

File: crates/ibc/src/applications/transfer/coin.rs [hunk 4]
```diff
@@ -59,24 +71,30 @@ where
         // a letter, a number or a separator ('/', ':', '.', '_' or '-').
         // Loosely copy the regex from here:
         // https://github.com/cosmos/cosmos-sdk/blob/v0.45.5/types/coin.go#L760-L762
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/coin.rs]:
```suggestion
        // https://github.com/cosmos/cosmos-sdk/blob/v0.47.5/types/coin.go#L838-L840
```

File: crates/ibc/src/applications/transfer/coin.rs [hunk 5]
```diff
@@ -119,53 +137,59 @@ impl<D: Display> Display for Coin<D> {
...
-            assert_eq!(coin.amount, 0u64.into());
-        }
+    use super::*;
 
-        {
-            // `!` is not allowed
-            let res = RawCoin::from_str("0x!");
-            assert!(res.is_err());
-        }
+    #[rstest]
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/coin.rs:135]: Good to include a test for checking lower/uppers case sensitivity.

File: crates/ibc/src/applications/transfer/coin.rs [hunk 6]
```diff
@@ -59,24 +71,30 @@ where
...
+        let (amount, denom) = coin_str
+            .chars()
+            .position(|x| !x.is_numeric())
+            .map(|index| coin_str.split_at(index))
+            .filter(|(amount, denom)| !amount.is_empty() && !denom.is_empty())
+            .filter(|(_, denom)| {
+                denom.chars().all(|x| {
+                    matches!(x, 'a'..='z' | 'A'..='Z' | '0'..='9' | '/' | ':' | '.' | '_' | '-')
+                })
+            })
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/coin.rs]: Here we can simplify the implementation and use the same approach as [here](https://github.com/cosmos/ibc-rs/blob/dc8e4161dd5e00a2031cf3feb58dad1ed7e26bdb/crates/ibc/src/core/ics24_host/identifier/validate.rs#L22-L31) by defnining e.g. `VALID_DENOM_CHARACTERS` , after which the `!denom.is_empty()` check won’t be needed anymore too. Even we might be able to decouple checks and filter of the amount from the denom. Appreciate a further look on this.

Something like this:

```rust
const VALID_DENOM_CHARACTERS: &str = "/:._-";

let (amount, denom) = coin_str
            .chars()
            .position(|x| !x.is_numeric())
            .map(|index| coin_str.split_at(index))
            .filter(|(amount, _)| !amount.is_empty())
            .filter(|(_, denom)| {
                denom
                    .chars()
                    .all(|x| x.is_alphanumeric() || VALID_DENOM_CHARACTERS.contains(x))
            })
            .ok_or_else(|| TokenTransferError::InvalidCoin {
                coin: coin_str.to_string(),
            })?;
```

**Review by Farhad-Shabani — CHANGES_REQUESTED** [review]

**Review by Farhad-Shabani — APPROVED** [review]
Good job @rnbguy. Thanks 👌🏻

**Review by soareschen — APPROVED** [review]

**Pull request merged**